importlib_metadata/__init__.py

```diff
@@ -153,6 +155,43 @@ def __reduce__(self):
         )
 
 
+class EntryPoints(tuple):
+    """
+    A collection of EntryPoint objects, retrievable by name.
+    """
+
+    def __getitem__(self, name) -> EntryPoint:
+        try:
+            return next(ep for ep in self if ep.name == name)
+        except Exception:
+            raise KeyError(name)
+
+    @property
+    def names(self):
+        return set(ep.name for ep in self)
+
+
+class GroupedEntryPoints(tuple):
+    """
+    A collection of EntryPoint objects, retrievable by group.
+    """
+
+    def __getitem__(self, group) -> EntryPoints:
+        return EntryPoints(ep for ep in self if ep.group == group)
+
+    @property
+    def groups(self):
+        return set(ep.group for ep in self)
+
+    def get(self, group, default=None):
+        """
+        For backward compatibility, supply .get
+        """
+        msg = "GroupedEntryPoints.get is deprecated. Just use __getitem__."
+        warnings.warn(msg, DeprecationWarning)
```

Comment:
seems odd to immediately deprecate this when there's no way to write the code which is compatible with the existing versions (`__getitem__` doesn't work for old importlib.metadata or importlib_metadata)

this is going to lead to a bunch of people reporting DeprecationWarning and flake8 being able to do nothing about it

Reply:
There are a few things flake8 could do:

- Instead of using `.get`, construct a dictionary explicitly from items (`eps = {ep.name: ep for ep in entry_points()}`). That method will work on old and new importlib_metadata.
- Suppress the deprecation warning.
- (my preference) just bump the requirement to the new release of `importlib_metadata` and use `__getitem__`.

Reply:
in case it wasn't clear -- I'm asking you please do not immediately deprecate something that's currently in use.  it will cause me a lot of pain dealing with users that can't read.

Reply:
I've never had such a request before, so I don't have any prior experience from which to draw. Part of the reason I engaged with the flake8 project was because of its prominence and my desire to address proactively the impact of this change on that project. How long should a deprecation warning be suppressed? Just until a subsequent release? A fixed period of time? Would a PendingDeprecationWarning be better for the interim or is any warning as bad as any other?

Reply:
other things in cpython seem to go:

- introduce new thing and don't warn in old thing until N releases pass (usually until the latest one with the old way is unsupported)
- PendingDeprecationWarning (>= 1 release)
- DeprecationWarning (>= 1 release)
- removed

your recommendation of bumping the version of `importlib_metadata` is very undesirable as it would require *re-introducing* a dependency in python3.8+ -- just to hack around this

Reply:
In eb170ba, I've put a 60 day delay on the deprecation warnings. Will that do the trick?

Reply:
no it will not, flake8 releases about once a year

Reply:
I want this deprecation warning to go into Python 3.10. I agree there are some cases where deprecations happen much slower, but this case has a workaround that works across versions _and_ a backport (this library), so the motivations for a long deprecation period aren't present.

Reply:
In c9adbca, I've added an escape hatch for flake8.

Reply:
(amended with 71fd4a7, because the prior commit was broken)